AI Ship Feedback, Voting Progress, and other bug fixes

hridaya423 opened the pull request:
This PR adds AI generated ship feedback using the votes for a ship's payout and shows voting progress on ships by showing how many votes it has

Fixes #827 
and fixes #749 
and fixes #898 
and fixes #901 

Features implemented:
- Admin manual regenerate feedback
- Backfilling job/service & rake
- Auto AI feedback after ship payout, saves into DB
- Show current amount of votes the latest ship has
- Not double escape project descriptions
- Fix extra overflow because of tilt & waves

Migration is required for DB
  rails generate migration AddFeedbackToShipEvents feedback:text
  rails db:migrate
  
for backfill: rake ship_feedback:backfill

These changes were partially tested (Couldn't test all, because im on windows and ruby doesnt like windows i think) but i think it should all work

cskartikey commented on app/controllers/ship_events_controller.rb:
```diff
@@ -0,0 +1,56 @@
+class ShipEventsController < ApplicationController
```
This should be a service. I don't really see this should have a route. Besides, break the PRs down to two.

cskartikey commented on app/javascript/controllers/ship_feedback_controller.js:
```diff
@@ -0,0 +1,106 @@
+import { Controller } from "@hotwired/stimulus"
```
How does this look? Attach some screenshots/screenrecording.

cskartikey commented:
Do note that I'm unlikely to merge anything that touches voting. Shipsfeedback seems fine tho. Although, @3kh0 was working on it.

hridaya423 commented:
> Do note that I'm unlikely to merge anything that touches voting. Shipsfeedback seems fine tho. Although, @3kh0 was working on it. (Ask for their review on the new PR)

The change isn't really with voting though, its just showing no of votes on project's hang tight text. (Also do you recommend me making 1 pr for each issue or only just rebase the big feature, ship feedback, and let the rest be as enhancements/bug fixes as 1 pr)

hridaya423 commented on app/models/ship_event.rb:
```diff
@@ -28,6 +28,28 @@ class ShipEvent < ApplicationRecord
+    feedback.match(/\*\*Summary\*\*:\s*(.+?)(?=\n\*\*|\z)/m)&.[](1)&.strip
+  end
+
+  def vote_count
```
.where("created_at > ?", created_at) this counts votes after the ship event was created? so i believe that filter is already existent unless i got the logic wrong

hridaya423 commented:
Moved regenerate feedback logic to model from controller, moved the regen route & controller to admin, and heres how it looks: 
<img width="728" height="476" alt="image" src="https://github.com/user-attachments/assets/cf0f5e99-da60-4e30-a2a1-42497e4a3af5" /> All changes related to above done

Spacexplorer11 commented:
Which model does the feedback use? ai.hackclub.com?

hridaya423 commented:
yep

cskartikey commented:
going to merge this tomorrow

hridaya423 commented:
@cskartikey hi

hridaya423 commented on app/views/projects/my_projects.html.erb:
```diff
@@ -42,7 +42,14 @@
-                    <p class="text-red-400">Your latest ship is being voted on. Hang tight - you'll get your payout soon!</p>
+                  <% remaining = current_user.remaining_votes_to_ship %>
+                  <% if remaining > 0 %>
+                    <p class="text-red-400">Your latest payout is on hold — you need <%= pluralize(remaining, 'more vote') %> since your last ship to release your shells. <a href="<%= new_vote_path %>" class="underline text-blue-700 hover:text-blue-800">Start voting</a>.</p>
```
what sounds better to you

cskartikey replied:
whatever is the latest. we're not doing escrow anymore
